src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -395,9 +395,17 @@ HTTPProxy::HTTPProxy() : _client(K23SIClientConfig()) {
 }
 
 seastar::future<> HTTPProxy::gracefulStop() {
-    _txns.clear();
-
-    return seastar::make_ready_future();
+    std::vector<seastar::future<>> futs;
+    for (auto& [ts, txn]: _txns) {
+        futs.push_back(txn.handle.kill());
+    }
+    return seastar::when_all_succeed(futs.begin(), futs.end()).discard_result()
+    .then([this] {
+        _txns.clear();
+    })
+    .then([this] {
+        return _client.gracefulStop();
```

Comment:
What happens if seastar calls `K23SIClient::gracefulStop()` before calling HttpProxy `gracefulStop`? Then _client gracefulstop will be called twice.

Reply:
We are not using the K23SIClient as an applet, so gracefulStop is not called for it automatically. Instead the HTTPProxy owns the client directly (it is a member variable) so it is responsible for starting and stopping it.

Reply:
Yes forgot that it's just client api not applet.